### ISSUE TITLE
REAL dependency inversion: Support internal dispatch by using a pair of traits

### ISSUE THREAD

**Issue by audunhalland** [audunhalland/entrait#8]
_TL;DR: The crate that defines an interface should not need to be the crate that implements it. The way entrait is designed, using delegating blanket implementations, requires some more delegation "magic"._

The entrait-for-trait feature `#[entrait(delegate_by = Borrow)]` only works for leaf dependencies.

Here is an idea for how to do something similar, but _without_ exiting the `Impl` layer.

We have some API that we want to potentially implement (downstream) in different dynamic ways:

```rust
pub trait Facade {
    fn foo(&self) -> i32;
}
```

We want to implement this trait for `Impl<T>` so it can be used as a dependency. But that, by definition, is the only implementation. We want to have a delegation through dynamic dispatch to reach the final destination. Since that trait is already implemented for `Impl<T>`, _we need another trait!_ We can use the entrait syntax to generate a trait from a trait 😬  Let's call the new trait `Backend`:

```rust
#[entrait(pub Backend)]
pub trait Facade {
    fn foo(&self) -> i32;
}
```

The generated trait is generic, and has two receivers:

```rust
pub trait Backend<T>: 'static {
    fn foo(&self, _impl: &entrait::Impl<T>) -> i32;
}
```

The generated delegation looks like this:
```rust
impl<T: 'static> Facade for Impl<T>
where
    T: core::borrow::Borrow<dyn Backend<T>>,
{
    fn foo(&self) -> i32 {
        self.borrow().foo(self)
    }
}
```

Now the application has to implement `Borrow<dyn Backend<Self>>`. This is the manual part.

To define an implementation of `Backend<T>`, we can write the following module:

```rust
pub struct MyImpl;

#[entrait_impl(some_crate::Backend for MyImpl)] <--- proposed syntax
#[entrait_impl(dyn some_crate::Backend for MyImpl)] <--- dyn version
mod my_impl {
    fn foo(deps: &impl Whatever) -> i32 {
         42
    }
}
```

The functions inside this module need to match the backend interface. I think the compile errors will be good enough. The implementation is of course auto-generated:

```rust
pub struct MyImpl;

impl<T> Backend<T> for MyImpl
where
    Impl<T>: Whatever,
{
    fn foo(&self, _impl: &Impl<T>) -> i32 {
        foo(_impl)
    }
}
```

The Borrow part of the application could either be a Box or some enum. Avoiding allocation could look like this:

```rust
impl Borrow<dyn facade::Backend<Self>> for App {
    fn borrow(&self) -> &dyn Backend<Self> {
        match &self.facade_impl {
            AppFacadeImpl::My(my_impl) => my_impl,
            AppFacadeImpl::Other(other_impl) => other_impl,
        }
    }
}
```

**Comment by audunhalland** [comment]
Note: This is also possible to do using static dispatch. It's much simpler, and the `Backend` trait is just a verbatim copy of the `Facade` trait. The `Facade` implementation very simply delegates with a `Self: Backend` bound. The only constraint is that there can only be one implementation of `Backend` for `Impl<T>`.

**Comment by audunhalland** [comment]
The only solution I found (so far) that allows static dispatch is this, which involves 3 traits:

```rust
use implementation::Impl;

pub trait Facade {
    fn foo(&self) -> i32;
}

pub trait FacadeImpl<T> {
    fn foo(__self: &Impl<T>) -> i32;
}

pub trait FacadeSelectImpl<T> {
    type Impl: FacadeImpl<T>;
}

// Delegation
impl<T> Facade for Impl<T> where T: FacadeSelectImpl<T> {
    fn foo(&self) -> i32 {
        <T as FacadeSelectImpl<T>>::Impl::foo(self)
    }
}
```

**Comment by audunhalland** [comment]
Solution exploration in this repo: https://github.com/audunhalland/entrait-dependency-inversion. There is a way to do it using one trait, but use custom smart pointers.

**Comment by audunhalland** [comment]
Implemented in #10 .

The final syntax is
```
#[entrait(TraitImpl, delegate_by = DelegationTrait)]
trait Trait {
    fn foo(&self);
}
```

I was not able to get zero-cost futures working when delegating using an alternative implementation of the original trait, .i.g. for `SomeImplRef<'s, T>(&'s Impl<T>)`. I wasn't able to make the resulting future implement `Send`.

The solution is to generate the `TraitImpl`, consisting of static methods:

```
trait TraitImpl<T> {
    fn foo(_impl: &Impl<T>);
}
```

The implementation block looks like:
```
#[entrait_impl]
mod my_impl {
    fn foo(deps: &impl Any) {
    }
    
    #[derive_impl(super::TraitImpl)]
    pub struct MyImpl;
}
```

We choose the implementation by writing:
```
impl DelegationTrait for App {
    type Target = my_impl::MyImpl;
}
```

**Comment by audunhalland** [comment]
Released in [0.4.4](https://github.com/audunhalland/entrait/commit/790d685534a99d70995906dcd3e9c28d878d4073)
